### ISSUE TITLE
[HOLD for payment 2023-11-21] [$250] Add error handling to readFilyAsync

### ISSUE THREAD

**Issue by luacmartins** [Expensify/App#24141]
# Problem

We added the ability to request money with a scanned receipt in [this PR](https://github.com/Expensify/App/pull/23770), but there's no error handling if we fail to read the file before sending the request

# Why is this important
Informs users that there was something wrong so it's not confusing for them.

# Solution
Add error handling
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0116eda5bd8fef8746</li>
        <li>Upwork Job ID: 1716521658283470848</li>
        <li>Last Price Increase: 2023-10-30</li>
<li>Automatic offers: </li>
<ul>
<li>DylanDylann | Contributor | 27491365</li>
</ul></li>
    </ul>
</details>

**Comment by luacmartins** [comment]
Not a priority. I'll try to work on it next week.

**Comment by luacmartins** [comment]
Not a priority. I'll try to work on it next week.

**Comment by luacmartins** [comment]
Not a priority

**Comment by luacmartins** [comment]
I'll try to get to this this week. We also have an additional consideration that sometimes selecting a file in iOS just fails.

**Comment by luacmartins** [comment]
Not a priority

**Comment by luacmartins** [comment]
I think we can clean up the method by adding `onSuccess` and `onFailure` callback params to it and then just calling `onFailure` in the catch instead of resolving the promise. That also cleans up the need to have promise chaining in other files. Gonna make this external.

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We added the ability to request money with a scanned receipt in https://github.com/Expensify/App/pull/23770, but there's no error handling if we fail to read the file before sending the request

### What is the root cause of that problem?
We added the ability to request money with a scanned receipt in https://github.com/Expensify/App/pull/23770, but there's no error handling if we fail to read the file before sending the request

### What changes do you think we should make in order to solve the problem?
As suggestion here https://github.com/Expensify/App/issues/24141#issuecomment-1775773774. We should add onSuccess and onFailure callback params to readFileAsync function. And we will call onFailure in the catch block instead of resolve promise. In onFailure callback, We can add logic to notify to user that the File upload has failed

### What alternative solutions did you explore? (Optional)

**Comment by luacmartins** [comment]
Waiting on @robertKozik to review proposals

**Comment by robertKozik** [comment]
@DylanDylann Can you provide more details about the implementation approach which you would execute? For now your proposal is very general

**Comment by luacmartins** [comment]
Here's an [example](https://github.com/Expensify/App/blob/8ec732744d2a65425b0fd26300a6625240feb607/src/pages/iou/steps/MoneyRequestConfirmPage.js#L104-L105). When the user uploaded a receipt and refreshed the page, we'd fail to load the file because the browser would get rid of the temp url. We have since fixed that issue, but I think we can clean it up with the solution I described [here](https://github.com/Expensify/App/issues/24141#issuecomment-1775773774)

**Comment by DylanDylann** [comment]
@luacmartins Thanks for your explanation. 
Currently, If user reload page, the browser gets rid of the temp URL and the App will redirect to Scan screen (upload receipt screen)
Do you mean when the browser gets rid of the temp URL, we need to redirect the user to Scan screen and **display a message like: "Please help to post receipt again"**

**Comment by luacmartins** [comment]
No, we should still reroute the user. This is just a refactor of the current solution.

**Comment by DylanDylann** [comment]
@robertKozik [Updated proposal](https://github.com/Expensify/App/issues/24141#issuecomment-1776282309) for more detail
cc @luacmartins

**Comment by robertKozik** [comment]
Great! Thanks @DylanDylann for updating your proposal. I think we are GTG
🎀 👀 🎀 C+ reviewed

**Comment by DylanDylann** [comment]
@robertKozik The PR is ready for review

**Comment by conorpendergrast** [comment]
Payouts due:

- [x] Contributor: $250 @DylanDylann [Offer](https://www.upwork.com/nx/wm/offer/27491365)

- Contributor+: @robertKozik N/A - contractor

Upwork job is [here](https://www.upwork.com/jobs/~0116eda5bd8fef8746).

**Comment by conorpendergrast** [comment]
This looks like a known uncaught error in a new feature rather than a bug itself, so I don't think we need the C+ checklist here. Closing out, reopen or comment if you disagree!
